mlonmcu/target/riscv/ara.py

```diff
@@ -0,0 +1,172 @@
+#
+# Copyright (c) 2022 TUM Department of Electrical and Computer Engineering.
+#
+# This file is part of MLonMCU.
+# See https://github.com/tum-ei-eda/mlonmcu.git for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""MLonMCU ARA Target definitions"""
+
+import os
+import re
+from pathlib import Path
+
+from mlonmcu.logging import get_logger
+from mlonmcu.feature.features import SUPPORTED_TVM_BACKENDS
+from mlonmcu.target.common import cli, execute
+from mlonmcu.target.metrics import Metrics
+from .riscv import RISCVTarget
+from .util import update_extensions
+import shutil
+
+logger = get_logger()
+
+class AraTarget(RISCVTarget):
+    """Target using a Pulpino-like VP running in the GVSOC simulator"""
+
+    FEATURES = RISCVTarget.FEATURES + ["log_instrs", "xpulp"]
+
+    DEFAULTS = {
+        **RISCVTarget.DEFAULTS,
+        "xlen": 64,
+        "abi": "lp64d",
+        "extensions": ["g", "c", "v"],  # TODO overwrite extensions elegantly
+        "nr_lanes": 4,
+        "vlen": 4096
+    }
+
+    REQUIRED = RISCVTarget.ARA_GCC_TOOLCHAIN_REQUIRED + [
+        "ara.apps_dir",
+        "ara.hardware_dir"
+    ]
+
+    def __init__(self, name="ara", features=None, config=None):
+        super().__init__(name, features=features, config=config)
+
+    @property
+    def ara_apps_dir(self):
+        return Path(self.config["ara.apps_dir"])
+
+    @property
+    def ara_hardware_dir(self):
+        return Path(self.config["ara.hardware_dir"])
+
+    @property
+    def Vara_tb_verilator(self):
```

Comment:
TODO: Recompile HW wehen VLEN for NR_LANES changes? (every time)

Reply:
I have added the function. But there are two problems. First is that the python code for this recompilation is very long. The second is that it takes some time to do the recompilation.

Reply:
I will try it out and report how to proceed

Reply:
Two tips here:
- You can always move that bunch of code to a separate file e.g. `ara_utils.py` to keep the `ara.py` file more clean
- If this might help to reduce the complexity of the verilator stuff: You could in theory also but a script into the `resources` subdir and call if from here using as subprocess.

Another question: Can’t we just use the Makefiles in the Ara repo for building the HW or does it lack any of the flexibility needed here?

Reply:
Good point. I will change it to use makefile later